components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt

```diff
@@ -0,0 +1,60 @@
+package net.corda.ledger.utxo.flow.impl.transaction.factory.impl
+
+import net.corda.ledger.common.data.transaction.WireTransaction
+import net.corda.ledger.utxo.data.transaction.UtxoLedgerTransactionImpl
+import net.corda.ledger.utxo.data.transaction.WrappedUtxoWireTransaction
+import net.corda.ledger.utxo.flow.impl.persistence.UtxoLedgerPersistenceService
+import net.corda.ledger.utxo.flow.impl.transaction.factory.UtxoLedgerTransactionFactory
+import net.corda.sandbox.type.UsedByFlow
+import net.corda.v5.application.serialization.SerializationService
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.base.exceptions.CordaRuntimeException
+import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction
+import net.corda.v5.serialization.SingletonSerializeAsToken
+import org.osgi.service.component.annotations.Activate
+import org.osgi.service.component.annotations.Component
+import org.osgi.service.component.annotations.Reference
+import org.osgi.service.component.annotations.ServiceScope
+
+@Component(
+    service = [UtxoLedgerTransactionFactory::class, UsedByFlow::class], scope = ServiceScope.PROTOTYPE
+)
+class UtxoLedgerTransactionFactoryImpl @Activate constructor(
+    @Reference(service = SerializationService::class)
+    private val serializationService: SerializationService,
+    @Reference(service = UtxoLedgerPersistenceService::class)
+    private val utxoLedgerPersistenceService: UtxoLedgerPersistenceService
+) : UtxoLedgerTransactionFactory, UsedByFlow, SingletonSerializeAsToken {
+
+    @Suspendable
+    override fun create(
+        wireTransaction: WireTransaction
+    ): UtxoLedgerTransaction {
+        val wrappedUtxoWireTransaction = WrappedUtxoWireTransaction(wireTransaction, serializationService)
+        val sourceTransactions =
+            (wrappedUtxoWireTransaction.inputStateRefs + wrappedUtxoWireTransaction.referenceStateRefs)
+                .map { it.transactionHash }
+                .toSet()
+                .associateWith { it ->
+                    utxoLedgerPersistenceService.find(it)?.outputStateAndRefs
+                        ?: throw (CordaRuntimeException("Could not find transaction $it when fetching input states."))
+                }
+        val inputStateAndRefs =
+            wrappedUtxoWireTransaction.inputStateRefs.map { it ->
+                sourceTransactions[it.transactionHash]?.get(it.index)
+                    ?: throw (CordaRuntimeException("Input state not found ${it.transactionHash} ${it.index}"))
```

Comment:
```suggestion
                    ?: throw (CordaRuntimeException("Input state not found ${it.ref}"))
```

Should use the standard `toString()` defined on the reference part, which will add a colon between the hash and index and ensures formatting consistency.

Comment also applies to line 51 as well.

Reply:
Thanks. I have updated them:
https://github.com/corda/corda-runtime-os/pull/2819/files#diff-e512a249344bb1ddd2949e337a93c0e25ef4c25b2af08a23a643bc1715c04122